src/bootstrap/src/core/build_steps/doc.rs

```diff
@@ -1292,8 +1306,19 @@ impl Step for Reference {
     }
 
     fn make_run(run: RunConfig<'_>) {
+        // Bump the stage to 2, because the reference requires an in-tree compiler.
+        // At the same time, since this step is enabled by default, we don't want `x doc` to fail
+        // in stage 1.
+        // FIXME: create a shared method on builder for auto-bumping, and print some warning when
+        // it happens.
+        let stage = if run.builder.config.is_explicit_stage() || run.builder.top_stage >= 2 {
+            run.builder.top_stage
+        } else {
+            2
+        };
```

Comment:
Remark: hmm, auto-bumping stages always feel a bit sad (because it can be quite surprising to the user), but I suppose this makes sense. Good call on the warning when it will happen.